### PR TITLE
[Google for Woo] Refactor Webview usage to use its own GoogleAdsWebView

### PR DESCRIPTION
**Pull request by hafizrahman**
Please do not merge until target is trunk.

This refactor is the implementation for the discussion in a previous PR:

https://github.com/woocommerce/woocommerce-android/pull/12013#discussion_r1683048918

Despite the big PR warning, most of this is just from the deletion of `ExitAwareWebView`.

### Description

We're adding a flag to tell `WPComWebView` to not authenticate if not needed, which is needed as a workaround for an issue related to Jetpack CP sites using Google for Woo features in the app.

The removes the need for having a separate `ExitAwareWebView`, which is now deleted.

Additionally, this PR also adds an event flow in `WPComWebView`. It will be needed later for Google for Woo tracking, so for now it just needs its code checked. The idea is to inject `SharedWebViewFlow` into any view model that needs it, and it can collect the `webViewEventFlow` to act on any event that it cares about ([example usage in the other PR](https://github.com/woocommerce/woocommerce-android/pull/12098/files#diff-1d2a6d541d2eb6081ac47608edf4914fb2b4292fff68a08e5ce8a1d24c9cb0a5R49))

### To test

Same testing instructions as in https://github.com/woocommerce/woocommerce-android/pull/11966

1. Test with a site on WordPress.com,
2. Use the Google for WooCommerce button in More Menu. The webview should automatically log you in and open the Google Ads page,
3. Switch to a site connected with application password/Jetpack CP,
4. Use the Google for WooCommerce button in More Menu. The webview should open a .org login form,
5. Login and ensure it opens the Google Ads page,
6. Close the webview and use the Google for WooCommerce button again,
7. Ensure that it opens the Google Ads page immediately, no login needed.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>GoogleAdsWebViewViewModel</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/227">19.7</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/wpcomwebview/SharedWebViewFlow.kt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import kotlinx.coroutines.flow.asSharedFlow
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * A singleton class responsible for managing web view events within the application's WebView
+ * (e.g: `WPComWebView`).
+ */
+@Singleton
+class SharedWebViewFlow @Inject constructor() {
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/wpcomwebview/SharedWebViewFlow.kt]: Good idea for this @hafizrahman, but sadly and given Android's quirks, this won't work as expected in some cases, I'll explain more below.

Given how this is being used, we assume that the previous fragment will stay active in the backstack while the user creates the campaign, but this is not true, if the Activity was re-created for some reason (the device was low on resources, and the app was on the background for example), only the top fragment (the last one on the backstack) will be re-created, the others won't be re-created until we pop the backstack, and this means that all events emitted during this time will be lost.

You can simulate this with the following (in the branch of the tracking):
1. Go to the developer options, and enable the option: `Don't keep activities`
2. Open the app, then start the campaign creation flow.
3. Switch to a different app, then switch back to the Woo app.
4. Perform any actions on the campaign creation, and notice that the events are not tracked.

You might say that this doesn't happen often, but in my experience it does happen quite often given the limited resources of Android devices.

File: WooCommerce/src/main/res/navigation/nav_graph_main.xml [hunk 2]
```diff
@@ -493,6 +493,10 @@
             android:defaultValue="@null"
             app:argType="string[]"
             app:nullable="true" />
+        <argument
+            android:name="skipAutoAuth"
+            android:defaultValue="false"
+            app:argType="boolean" />
```
Comment (hichamboushaba) [inline WooCommerce/src/main/res/navigation/nav_graph_main.xml]: Just a discussion regarding this flag, why do we need it? couldn't we just just check the `SelectedSite` insite the `WPComWebViewAuthenticator` and decide when to authenticate and when not?

**Review by hichamboushaba — CHANGES_REQUESTED** [review]
Thanks @hafizrahman for working on this.

As shared in the comments, there is a major issue with the approach used in the PR, and I'm afraid there is not an easy solution to achieve what you are trying while reusing the `WPComWebViewFragment`, I think the easiest solution for you is to have a specific fragment used only for the Google Ads campaign creation, the `WPComWebViewFragment` is intended for more simple usages.

As an example for how to approach this, you can check how we handled the Blaze campaign creation in the WebView, it's very similar to what you are trying to achieve here, and should provide a clear idea of how to it while reusing many parts of the `WebView` architecture in the app.
The code is available here: https://github.com/woocommerce/woocommerce-android/tree/e49ff61cbdd887b6726c9c8f07c8e7b28ab8017c/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze, the classes suffixed with `BlazeCampaignCreation`

(PS: the same approach is used also for some other screens, like the Theme preview where we needed some advanced handling)

Please let me know if I can be of any help.

**Comment by hafizrahman** [comment]
@hichamboushaba Thanks for the comment, that's an angle that I hadn't considered and I'm glad I got you to review it. I'll refactor using the example in `BlazeCampaignCreation-`.

**Comment by hafizrahman** [comment]
@hichamboushaba I added the refactor by creating a separate `GoogleAdsWebView` as you suggested. The code more or less looks the same as the previous iteration, without the flow. For tracking, the plan is just to do it directly inside `GoogleAdsWebViewViewModel`'s various functions.

> Just a discussion regarding this flag, why do we need it? couldn't we just just check the SelectedSite insite the WPComWebViewAuthenticator and decide when to authenticate and when not?

As for this, I'm creating a specific usecase `CanUseAutoLoginWebview` just to make it super clear what the criteria is, and to make things easier to test also. With this current refactor, it is only injected into `GoogleAdsWebViewViewModel`.

 Let me know what you think!

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/google/webview/GoogleAdsWebViewFragment.kt [hunk 3]
```diff
@@ -17,7 +17,7 @@ import com.woocommerce.android.viewmodel.MultiLiveEvent.Event.ExitWithResult
 import dagger.hilt.android.AndroidEntryPoint
 
 @AndroidEntryPoint
-class ExitAwareWebViewFragment : BaseFragment(R.layout.fragment_exitaware_webview) {
+class GoogleAdsWebViewFragment : BaseFragment(R.layout.fragment_googleads_webview) {
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/google/webview/GoogleAdsWebViewFragment.kt]: I don't think we need a layout file for this Fragment, it's using `Compose`.

I think this was a copy paste from the `WPComWebViewFragment`, and I mistakenly didn't remove this part when I converted it to Compose 🤦, so for your case now, you can remove this part and remove the file `R.layout.fragment_googleads_webview`

**Review by hichamboushaba — APPROVED** [review]
Nice work @hafizrahman, the code looks great, I left just one small remark regarding the usage of the layout file `R.layout.fragment_googleads_webview`, it's not blocking, but it would be great to remove it before merging the PR.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/google/webview/GoogleAdsWebViewFragment.kt (hunk 3 above)
Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/google/webview/GoogleAdsWebViewFragment.kt]: Thanks for catching that. Done https://github.com/woocommerce/woocommerce-android/pull/12097/commits/7b8c8b9727d65f01a5bd1eeebbe37d0b50e83c64

**Pull request merged**